utils/common/src/main/java/org/apache/brooklyn/util/repeat/Repeater.java

```diff
@@ -301,89 +332,96 @@ public Repeater limitTimeTo(Duration duration) {
     public boolean run() {
         return runKeepingError().getWithoutError();
     }
-    
+
     public void runRequiringTrue() {
         Stopwatch timer = Stopwatch.createStarted();
         ReferenceWithError<Boolean> result = runKeepingError();
         result.checkNoError();
-        if (!result.get()) 
+        if (!result.get()) {
             throw new IllegalStateException(description+" unsatisfied after "+Duration.of(timer));
+        }
     }
-    
+
     public ReferenceWithError<Boolean> runKeepingError() {
-        Preconditions.checkState(body != null, "repeat() method has not been called to set the body");
-        Preconditions.checkState(exitCondition != null, "until() method has not been called to set the exit condition");
-        Preconditions.checkState(delayOnIteration != null, "every() method (or other delaySupplier() / backoff() method) has not been called to set the loop delay");
+        Preconditions.checkNotNull(exitCondition, "until() method has not been called to set the exit condition");
+        Preconditions.checkNotNull(delayOnIteration, "every() method (or other delaySupplier() / backoff() method) has not been called to set the loop delay");
 
         boolean hasLoggedTransientException = false;
         Throwable lastError = null;
         int iterations = 0;
         CountdownTimer timer = timeLimit!=null ? CountdownTimer.newInstanceStarted(timeLimit) : CountdownTimer.newInstancePaused(Duration.PRACTICALLY_FOREVER);
 
-        while (true) {
-            Duration delayThisIteration = delayOnIteration.apply(iterations);
-            iterations++;
-
-            try {
-                body.call();
-            } catch (Throwable e) {
-                log.warn(description, e);
-                if (rethrowImmediatelyCondition.apply(e)) throw Exceptions.propagate(e);
-            }
-
-            boolean done = false;
-            try {
-                lastError = null;
-                done = exitCondition.call();
-                hasLoggedTransientException = false;
-            } catch (Throwable e) {
-                if (hasLoggedTransientException) {
-                    if (log.isDebugEnabled()) log.debug(description + " (repeated failure; excluding stacktrace): " + e);
-                } else {
-                    if (log.isDebugEnabled()) log.debug(description, e);
-                    hasLoggedTransientException = true;
+        try {
+            while (true) {
+                Duration delayThisIteration = delayOnIteration.apply(iterations);
+                iterations++;
+
+                Future<?> call = executor.submit(body);
+                try {
+                    call.get(delayThisIteration.toMilliseconds(), TimeUnit.MILLISECONDS);
```

Comment:
I think a separate time out is needed here, especially when used with `backoff`, but happy to go with this.